Architecture_Engine/Query/Distance.cs

```diff
@@ -0,0 +1,90 @@
+﻿/*
+ * This file is part of the Buildings and Habitats object Model (BHoM)
+ * Copyright (c) 2015 - 2021, the respective contributors. All rights reserved.
+ *
+ * Each contributor holds copyright over their respective contributions.
+ * The project versioning (Git) records all such contribution source information.
+ *                                           
+ *                                                                              
+ * The BHoM is free software: you can redistribute it and/or modify         
+ * it under the terms of the GNU Lesser General Public License as published by  
+ * the Free Software Foundation, either version 3.0 of the License, or          
+ * (at your option) any later version.                                          
+ *                                                                              
+ * The BHoM is distributed in the hope that it will be useful,              
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of               
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the                 
+ * GNU Lesser General Public License for more details.                          
+ *                                                                            
+ * You should have received a copy of the GNU Lesser General Public License     
+ * along with this code. If not, see <https://www.gnu.org/licenses/lgpl-3.0.html>.      
+ */
+
+using BH.oM.Architecture.BuildersWork;
+using BH.oM.Architecture.Elements;
+using BH.oM.Geometry;
+using System.Collections.Generic;
+using System.Linq;
+using BH.oM.Spatial.ShapeProfiles;
+using BH.Engine.Geometry;
+using BH.oM.Geometry.CoordinateSystem;
+
+namespace BH.Engine.Architecture
+{
+    public static partial class Query
+    {
+        /***************************************************/
+        /**** Public Methods                            ****/
+        /***************************************************/
+
+        public static double Distance(this Opening opening1, Opening opening2, double maxExpectedDistance = double.NaN, double distanceTolerance = Tolerance.Distance, double angleTolerance = Tolerance.Angle)
+        {
+
+            if (opening1 == null || opening2 == null)
+            {
+                BH.Engine.Reflection.Compute.RecordError("Opening is null");
+                return double.NaN;
+            }
+
+            double minDistance = double.MaxValue;
+
+            Point centerPoint2 = opening2.CoordinateSystem.Origin;
+            Plane plane2 = new Plane { Origin = centerPoint2, Normal = opening2.CoordinateSystem.Z };
+
+            if (!opening1.IsValidForDistanceCheck(opening2, maxExpectedDistance, distanceTolerance, angleTolerance)) 
+            {
+                BH.Engine.Reflection.Compute.RecordError("Distance can't be calculated");
+                return double.NaN;
+            }
```

Comment:
I would move this `if` statement before line 49 and take the code below outside the `else` statement. By doing that you will first have method applicability checks, with the actual query following afterwards.

So it will be:
1. check if the openings are not null
2. check if the openings are valid for distance check
3. if both above are OK then start declaring variables and calculating stuff

Besides that, as above: please make the error message slightly more informative (why can't the distance be calculated?) - using full stops at the end of sentences will earn you an extra CI/CD credit (@FraserGreenroyd 😃)